src/app/reporting/Engine.cpp

```diff
@@ -290,21 +291,92 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeReportIBs(ReportDataMessage::Bu
 }
 
+CHIP_ERROR Engine::RemoveUnaccessableEventPaths(TLV::TLVWriter & aWriter, bool & aItemRemoved, ReadHandler * apReadHandler)
+{
+    CHIP_ERROR err                                                      = CHIP_NO_ERROR;
+    ObjectList<EventPathParams> * deletedEventPathList                  = nullptr;
+    ObjectList<EventPathParams> ** previousIteratedEventPathListNextPtr = &(apReadHandler->mpEventPathList);
+
+    aItemRemoved = false;
+    for (auto current = apReadHandler->mpEventPathList; current != nullptr;)
+    {
+        if (current->mValue.HasEventWildcard())
+        {
+            current = current->mpNext;
+            continue;
+        }
+
+        Access::RequestPath requestPath{ .cluster = current->mValue.mClusterId, .endpoint = current->mValue.mEndpointId };
+        ConcreteEventPath path;
+        path.mClusterId                    = current->mValue.mClusterId;
+        path.mEndpointId                   = current->mValue.mEndpointId;
+        path.mEventId                      = current->mValue.mEventId;
+        Access::Privilege requestPrivilege = RequiredPrivilege::ForReadEvent(path);
+
+        err = Access::GetAccessControl().Check(apReadHandler->GetSubjectDescriptor(), requestPath, requestPrivilege);
+        if (err != CHIP_ERROR_ACCESS_DENIED)
+        {
+            SuccessOrExit(err);
+
+            // Proceed to the next item
+            previousIteratedEventPathListNextPtr = &(current->mpNext);
+            current                              = current->mpNext;
+        }
+        else
+        {
+            TLV::TLVWriter checkpoint = aWriter;
+            err                       = EventReportIB::ConstructEventStatusIB(aWriter, path,
+                                                        StatusIB(Protocols::InteractionModel::Status::UnsupportedAccess));
+            if (err != CHIP_NO_ERROR)
+            {
+                aWriter = checkpoint;
+                break;
+            }
+
+            ChipLogDetail(InteractionModel,
+                          "Removed event path (%d, " ChipLogFormatMEI ", " ChipLogFormatMEI ") due to insufficient privilege",
+                          current->mValue.mEndpointId, ChipLogValueMEI(current->mValue.mClusterId),
+                          ChipLogValueMEI(current->mValue.mEventId));
+
+            auto next = current->mpNext;
+
+            // Detach current item from the interested path list
+            *previousIteratedEventPathListNextPtr = next;
+
+            // Append current item to the deleted event path list
+            current->mpNext      = deletedEventPathList;
+            deletedEventPathList = current;
+
+            // proceed to the next item
+            current = next;
+        }
+    }
+
+exit:
+    // Always release the items in deletedEventPathList to avoid memory leak
+    if (deletedEventPathList != nullptr)
+    {
+        aItemRemoved = true;
+        InteractionModelEngine::GetInstance()->ReleaseEventPathList(deletedEventPathList);
```

Comment:
So hold on.  If we permanently remove the paths here, then the ACLs change to allow those paths, isn't that a problem?  The spec is really not clear on where the path set comes from for non-priming reads on a subscription...  @CamWms @mrjerryjohns what is the expected behavior there?

Reply:
I think we should not permanently remove path here.

Reply:
The spec states:

> Else if reading the event in the path requires a privilege that is not granted to access the cluster in the path, an EventStatusIB SHALL be generated with the UNSUPPORTED_ACCESS Status Code.

> If an EventStatusIB was generated, the path SHALL be discarded.

I read that to mean that the path is discarded _permanently_ for the duration of the subscription.

Reply:
I think we need to get the spec sorted.  It's pretending like there are no reads after the priming read....

Discarding permanently would be OK if that were clearly stated.  But then we should do it for attributes too, right?

Reply:
during the subscription, if the path is permanently removed during the priming stage since the event path don't have enough privilege, then later that path has privilege, how can we handle this situation?